# Optimizing Code: Common Books
Here's the code your coworker wrote to find the common book ids in `books_published_last_two_years.txt` and `all_coding_books.txt` to obtain a list of recent coding books.

The goal is to find the **coding books** among the recently published books based on identifier.<br>
Result should be  `recent_coding_books`.

In [1]:
import time
import pandas as pd
import numpy as np

In [3]:
with open('books_published_last_two_years.txt') as f:
    recent_books = f.read().split('\n')
    
with open('all_coding_books.txt') as f:
    coding_books = f.read().split('\n')

In [4]:
start = time.time()
recent_coding_books = []

for book in recent_books:
    if book in coding_books:
        recent_coding_books.append(book)

print(len(recent_coding_books))
print('Duration: {} seconds'.format(time.time() - start))

96
Duration: 17.34901738166809 seconds


### Tip #1: Use vector operations over loops when possible

Use numpy's `intersect1d` method to get the intersection of the `recent_books` and `coding_books` arrays.<br>
Numpy's documentation is https://numpy.org/doc/stable/reference/generated/numpy.intersect1d.html

In [9]:
start = time.time()
recent_coding_books = np.intersect1d(coding_books, recent_books)  # TODO: compute intersection of lists
print(len(recent_coding_books))
print('Duration: {} seconds'.format(time.time() - start))

96
Duration: 0.033902645111083984 seconds


As we can observe, the running time of intersection computation is reduced several times.

### Tip #2: Know your data structures and which methods are faster
Use the set's `intersection` method to get the common elements in `recent_books` and `coding_books`.

First, we need to *convert list to sets*. Then, apply `.intersect` in-built function.

In [11]:
start = time.time()
recent_coding_books =set(coding_books).intersection(set(recent_books))# TODO: compute intersection of lists
print(len(recent_coding_books))
print('Duration: {} seconds'.format(time.time() - start))

96
Duration: 0.01383829116821289 seconds


Now, we managed to optimise the code to run within **0.014 seconds**.